Generate real merge conflicts on template update

grst opened the pull request:
Close #175 

- [x]  make sure that in the future, after creating a repo, a first commit "initialize from template" is created.
- [x] ignore certain files (i.e. example files like basic.py). This should already be defined in cruft, but we need to respect it.
- [x] fix/update tests
- [x] suggest to use uv in the README (#370)

Zethson commented on scripts/src/scverse_template_scripts/cruft_prs.py:
```diff
@@ -129,95 +133,208 @@ def matches_current_version(self, pr: PullRequest) -> bool:
+            "https://github.com/scverse/cookiecutter-scverse",
+            f"--checkout={tag_name}",
+            "--no-input",
+            f"--extra-context-file={cookiecutter_config}",
```
What happens if the `cookiecutter_config` doesn't have all variables that the template has? IIRC it just takes the defaults/options from the template cookiecutter.json and doesn't break, right? It's been a while...

grst replied:
It takes the defaults from the cookiecutter.json.

Zethson reviewed (approved):
Great! This all looks sane to me.

I'll add a commit where I improve some nitpicky stuff. Not gonna bother you with that.

flying-sheep commented on scripts/tests/test_cruft.py:
I’ll look at the tests once we decided about the responsemock thing.

flying-sheep reviewed (changes requested):
Awesome functionality, some questions about the safeguards you removed.

grst commented on scripts/pyproject.toml:
```diff
@@ -86,14 +84,15 @@ lint.select = [
-  "S311",   # we don’t need cryptographically secure RNG
-  "S603",   # subprocess with shell=False should be allowed
+  "ISC001",  # conflicts with formatter
+  "PLR0913", # too many arguments in function definition
+  "PLR0915", # too many statements
```
I'd tend to agree, but since this is not a publicly-facing API, I didn't bother to refactor accordingly.

flying-sheep commented on scripts/pyproject.toml:
```diff
@@ -110,8 +109,6 @@ column_width = 120
 [tool.pytest.ini_options]
 addopts = [
   "--import-mode=importlib",
-  "--disable-socket",
-  "-presponsemock",
```
how do you guarantee that the real APIs are never called with too-high permissions and a mistake here wreaks havoc on our production data?

grst replied:
What if there were a separate github account, e.g. `scverse-bot-test`?

grst replied:
yeah that could work. Still one could be using a different token by mistake. 
But I guess if one stores it in a variable `SCVERSE_BOT_GH_TOKEN_READONLY` or something it should be quite clear.

grst replied:
So shall we go with the readonly token? I'm fine either way, but if you prefer to go with "no API calls" I'd probably strip down testing to a bare minimum as I don't have the time (and don't think it's worth the effort in this case) to spend a lot more time on this.

flying-sheep replied:
yeah, ideally in a way that makes it easy to locally provide the token, e.g. via env variable

flying-sheep commented on scripts/src/scverse_template_scripts/_log.py:
```diff
@@ -0,0 +1,13 @@
+    from rich.traceback import install
+
+    basicConfig(level="INFO", handlers=[RichHandler()])
+    install(show_locals=True)
```
I see! Well, if it was there before, it’s probably fine.

Zethson replied:
Oh, I missed that. Let's remove it, please. Also makes copy and pasting annoying.

flying-sheep replied:
```suggestion
```

grst commented:
@flying-sheep, I refactored the code a little that separate units can be tested and added unit tests for all steps that can be tested with a readonly github token. 

Btw, here's a test PR created by the script: https://github.com/icbi-lab/infercnvpy/pull/160

flying-sheep commented:
Hi! Unfortunately I got sick and can’t really rely on thinking straight. I think I should be good around Thursday. You could also ask someone else to review instead of course!

grst commented:
take your time, I don't think it's that urgent. 
I'll also be away on a long weekend, so maybe we can try to get it in early next week.

flying-sheep commented on scripts/tests/test_cruft.py:
```diff
@@ -1,75 +1,158 @@
+    assert config["context"]["cookiecutter"]["project_name"] == "cookiecutter-scverse-instance"
+
+
+def test_apply_update(cookiecutter_scverse_instance_cloned_repo: Repo, current_repo_path: Path, tmp_path: Path) -> None:
```
what does this test? just that it doesn’tr crash?

flying-sheep reviewed (approved):
OK, looks great! Only the one question above.

I made a few small changes:
- there were some fixes that I could make by turning type checking on, please check out afb40fc20a78725e63e1a6d8309f43a71c4b4189
- I added types to the test folder and could then shorten fixture names which makes the test code more concise
- I renamed `n_retries` to be more specific to what the “5 minutes” refer to

Otherwise, please upload the readonly token to 1Password so it’s easy to grab. (Maybe 1Password supports shareable links that work for people allowed to access the resource? Then you could add that to the comments)

grst commented:
Thanks for the review+fixes! I added the token to 1password, but didn't find a way to generate a link that wouldn't expose the password publicly.

grst commented:
TODO
 - [ ] the branch name doesn't encode the version number anymore. So PRs will be skipped if they already exist even when from an earlier version. That the branch name doesn't encode the version anymore is done on purpose - if merge conflicts got fixed on that branch at some point, future template updates should be based on that fixed version. So we need to find an alternative for skipping PRs here, or just not skip them at all, but just push updates to the same branch (this is also what pre-commit is doing)
 - [ ] disable dry run again (enabled to be able to make test releases without actually sending out PRs)

grst commented:
Ok, dryrun was successful now: https://github.com/scverse/cookiecutter-scverse/actions/runs/14188981598/job/39749362236

Shall we risk it and merge this + make a release?

grst commented:
Yes, there are a couple of repos, that have been deleted or archived that will fail. 
Maybe there are others that fail for weird edge-cases, but if we can get 95% I'm happy for now.

grst commented:
Ok, I'll hit the button at some point this week, but I want to make sure I have time to intervene and fix things if something goes wrong.
